pkg/transport/generic_transport_controller.go

```diff
@@ -0,0 +1,512 @@
+	return c.updateWrappedObjectsAndFinalizer(ctx, placementDecision)
+}
+
+// isObjectBeingDeleted is a helper function to check if object is being deleted.
+func isObjectBeingDeleted(object metav1.Object) bool {
+	return !object.GetDeletionTimestamp().IsZero()
+}
+
+func (c *genericTransportController) deleteWrappedObjectsAndFinalizer(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) error {
+	for _, destination := range placementDecision.Spec.Destinations { // TODO need to revisit the destination struct and see how to use it properly
+		if err := c.transportClient.Resource(c.wrappedObjectGVR).Namespace(destination.ClusterId).Delete(ctx, fmt.Sprintf("%s-%s", placementDecision.GetName(), c.wdsName),
+			metav1.DeleteOptions{}); err != nil { // wrapped object name is in the format (PlacementDecision.GetName()-WdsName). see updateWrappedObject func for explanation.
+			if !strings.HasSuffix(err.Error(), notFoundErrorSuffix) { // if object is already not there, we do not report an error cause desired state was achieved.
+				return fmt.Errorf("failed to delete wrapped object '%s' in destination WEC with namespace '%s' - %w", fmt.Sprintf("%s-%s", placementDecision.GetName(),
+					c.wdsName), destination.ClusterId, err)
+			}
+		}
+	}
+
+	if err := c.removeFinalizerFromPlacementDecision(ctx, placementDecision); err != nil {
+		return fmt.Errorf("failed to remove finalizer from PlacementDecision object '%s' - %w", placementDecision.GetName(), err)
+	}
+
+	return nil
+}
+
+func (c *genericTransportController) removeFinalizerFromPlacementDecision(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) error {
+	return c.updatePlacementDecision(ctx, placementDecision, func(placementDecision *v1alpha1.PlacementDecision) bool {
+		return removeFinalizer(placementDecision, transportFinalizer)
+	})
+}
+
+func (c *genericTransportController) addFinalizerToPlacementDecision(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) error {
+	return c.updatePlacementDecision(ctx, placementDecision, func(placementDecision *v1alpha1.PlacementDecision) bool {
+		return addFinalizer(placementDecision, transportFinalizer)
+	})
+}
+
+func (c *genericTransportController) updateWrappedObjectsAndFinalizer(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) error {
+	if err := c.addFinalizerToPlacementDecision(ctx, placementDecision); err != nil {
+		return fmt.Errorf("failed to add finalizer to PlacementDecision object '%s' - %w", placementDecision.GetName(), err)
+	}
+
+	objectsToPropagate, err := c.getObjectsFromWDS(ctx, placementDecision)
+	if err != nil {
+		return fmt.Errorf("failed to get objects to propagate to WECs from PlacementDecision object '%s' - %w", placementDecision.GetName(), err)
+	}
+
+	wrappedObject, err := convertObjectToUnstructured(c.transport.WrapObjects(objectsToPropagate))
+	if err != nil {
+		return fmt.Errorf("failed to wrap objects to a single wrapped object - %w", err)
```

Comment:
The failure here is not in the wrapping but just in the conversion to unstructured.

Reply:
fixed